Switch tool document cache to diskcache

mvdbeek opened the pull request:
This should be more portable across different Python installations.
If the fallback dbm.dumb module was used (instead of dbm.gnu or dbm/ndbm) startup could become very, very slow. diskcache uses SQLite and plain files to store the cache. With this being more portable we should be able to store this on CVMFS and have other Galaxy instances be able to consume the cache.

mvdbeek committed:
Switch tool document cache to diskcache

This should be more portable across different python installations.
If the fallback dbm.dumb module was used (instead of dbm.gnu or
dbm/ndbm) startup could become very, very slow. diskcache uses sqlite
and plain files to store the cache. With this being more portable we
should be able to store this on cvmfs and have other Galaxy instances be
able to consume the cache.

nsoranzo commented on lib/galaxy/dependencies/pipfiles/default/Pipfile:
```diff
@@ -58,6 +58,7 @@ WebOb = "*"
 Mako = "*"
 Babel = "*"
 Beaker = "*"
+diskcache = "*"
```
Can you also remove `dogpile.cache` from here and `lib/galaxy/dependencies/pipfiles/default/pinned-requirements` (assuming it's not needed by other packages)?

mvdbeek replied:
Sure!

mvdbeek replied:
Actually, I think it might be, we've had dog.pile in our dependencies before I started using it for the document cache (https://github.com/galaxyproject/galaxy/blob/release_20.01/lib/galaxy/dependencies/pipfiles/default/pinned-requirements.txt#L55)

nsoranzo replied:
OK, but you can still remove it from this file (`Pipfile`), which lists our explicit dependencies.

mvdbeek replied:
I'm a bit confused, it's not in this file, is it ?

nsoranzo replied:
Ah, it should have been there, but you're right it's not, so nothing to do!

nsoranzo replied:
This is not used any more, should be replaced by `sqlitedict`.

nsoranzo commented on lib/galaxy/tools/cache.py:
```diff
@@ -1,22 +1,15 @@
-    NO_VALUE,
-)
-from dogpile.cache.proxy import ProxyBackend
+from diskcache import Cache, JSONDisk
```
```suggestion
from diskcache import (
    Cache,
    JSONDisk
)
```

nsoranzo commented on lib/galaxy/tools/cache.py:
```diff
@@ -25,65 +18,42 @@
 CURRENT_TOOL_CACHE_VERSION = 2
```
Should we increase (or reset) this value? We now cache `document` instead of `metadata`.

mvdbeek replied:
```suggestion
CURRENT_TOOL_CACHE_VERSION = 1
```

mvdbeek committed:
Reset tool cache version, reorder imports

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo committed:
Move tool creation logic from ``cache.py``

to ``__init__.py``

mvdbeek committed:
Merge pull request #14 from nsoranzo/diskcache_cache_move_logic

Move tool creation logic from ``cache.py``

mvdbeek committed:
Use Index instead of Cache

Index doesn't expire items and has a dictionary-like interface.

mvdbeek committed:
Use plain JSON files to stored expanded tool sources

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -289,26 +294,40 @@ def tools_by_id(self):
-            self.cache_regions[tool_cache_data_dir] = create_cache_region(tool_cache_data_dir)
+            try:
+                with open(os.path.join(tool_cache_data_dir, 'cache.json')) as cache_file:
+                    self.cache_regions[tool_cache_data_dir] = json.load(cache_file)
```
Can this JSON file become too large to always have it loaded in memory as part of `app.toolbox` ? I liked having a database-backed cache :pensive:

mvdbeek replied:
Yeah, I'm sorry, I should have put this back into WIP. I have another variant that uses hdf5py that could work, but ... looks like that also has problems with multiple readers and writers. I am now considering writing a single JSON file per tool or debugging the diskcache/sqlite approach ... but I don't have a good lead on how to do that.

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -566,7 +585,7 @@ def remove_from_cache(self):
         if source_path:
             for region in self.app.toolbox.cache_regions.values():
-                region.delete(source_path)
+                region.pop(source_path, None)
```
I think we need to call `persist_cache_region()` here.

mvdbeek replied:
Yep, thanks! I've moved that into the caller so that we just need to persist once per set of removed tools.

mvdbeek committed:
Implement a HDF5 based cache

This isn't that useful since we can only have a single writer process.
If just reading a file that's fine, but multiple writers won't work
this way, so without a lot of additional work synachronizing cache
writes we can't really use this.

mvdbeek committed:
Try sqlite dict, didn't find a way to start multiple processes

The thread synchronization and re-raising logic seem to make this
impossible.

mvdbeek committed:
Write to process-local sqlite database

This should be a fairly robust way to work around multiple writers.

nsoranzo commented on lib/galaxy/tools/cache.py:
```diff
@@ -1,89 +1,89 @@
+            self._cache = SqliteDict(self.cache_file, flag='r', encode=encoder, decode=decoder, autocommit=False)
+
+    def set(self, config_file, tool_source):
+        self.make_writable()
```
I think it would be better to reopen the cache file inside a `with` statement otherwise other threads/processes would not be able to write to it.

mvdbeek replied:
So the goal of that with statement is to avoid re-assigning `self._cache` here ?

nsoranzo replied:
I realise now that's a bad idea since instantiating `SqliteDict` creates a thread to serialise access to the connection.

Why not have it always open for read and write?

mvdbeek replied:
The idea is that `make_writable` creates a tempfile next to the actual cache where the actual writing happens. When that is done we just the file back into place and re-open in read-only mode (read-only mode is technically not necessary since we call `make_writable` for every action that updates the database). This way multiple processes can write to the database at the same time. If we don't do this we get `database locked` exceptions when you start up multiple separate Galaxy processes at the same time (if the cache database is empty, if it's fully populated we don't ever call any of the update methods) -- we could place a lock around the database file, and serialize access that way, but we haven't had good success with file locking on the Anvil / GCP VMs.

mvdbeek committed:
Don't use RenamedTemproaryFile

And use a regular NamedTemporaryFile, deleting it once
the ToolDocumentCache instance goes out of scope.
Otherwise we may be moving an old cache over a newer cache when a
ToolDocumentCache instance goes out of scope.
I think it's better to explcitly call `persist` for any moving to
happen.

mvdbeek committed:
Persist cache when tools removed from cache

nsoranzo commented on lib/galaxy/tools/cache.py:
```diff
@@ -1,89 +1,96 @@
+        tool_document = self._cache.get(config_file)
+        if not tool_document:
+            return None
+        if tool_document.get('tool_cache_version', 0) != CURRENT_TOOL_CACHE_VERSION:
```
```suggestion
        if tool_document.get('tool_cache_version') != CURRENT_TOOL_CACHE_VERSION:
```

mvdbeek committed:
Don't set default value for cache version comparison

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on lib/galaxy/dependencies/pipfiles/default/pinned-requirements.txt:
```diff
@@ -49,6 +49,7 @@ defusedxml==0.6.0 ; python_version >= '3.0'
 deprecated==1.2.10
 deprecation==2.1.0
 dictobj==0.4
+diskcache==5.0.2
```
Not used any more.

mvdbeek committed:
Drop diskcache requirement, add sqlitedict
